### PR TITLE
[proposal] Add section about how you become a committer

### PR DESCRIPTION
**Pull request by flub**
This is just a proposal.  Feel free to discuss the intent or wording or add whatever thoughts you might have.

### CODE REVIEW

File: CONTRIBUTING.rst [hunk 1]
```diff
@@ -236,3 +236,15 @@ but here is a simple overview:
     base: features        # if it's a feature
 
 
+Joining the Development Team
+----------------------------
+
+Anyone who has successfully seen through a pull request which did not
+require any extra work from the development team to merge will
+themselves gain commit access (if we forget please send a friendly
+reminder).  This does not mean your workflow to contribute changes,
```
Comment (The-Compiler) [inline CONTRIBUTING.rst]: I had to re-read this a couple of times, parsed it as "(This does not mean your workflow to) (contribute changes)" (contributing changes, not changing a workflow) :wink:

**Comment by The-Compiler** [comment]
So I'm still not so sure about the potential of abuse. Sure, hopefully everything will go well, but if Mallory wants to push some malicious code to pytest, they can now trivially do so.

Would someone notice? Is anyone actually looking at the diff each time they pull? I doubt it :wink:

Also, please consider that people might not actually want those permissions, especially beginners or people insecure with how to use git might be scared and put-off by having the possibility of "messing things up" when accidentally pushing to a wrong place.

File: CONTRIBUTING.rst (hunk 1 above)
Comment (obestwalter) [inline CONTRIBUTING.rst]: @The-Compiler same here.

**Comment by nicoddemus** [comment]
What should we do about this guys?

**Comment by obestwalter** [comment]
I dunno. There are 16 people who could push malicious code to pytest already. I could push malicious code to tox. Why do we trust those people and which criteria do we have to invite people into our "circle of trust"? Or should we make it impossible for everyone to push code without another pair of eyes having checked the changes? Would that be possible with the permissions Github provides?

**Comment by The-Compiler** [comment]
> Or should we make it impossible for everyone to push code without another pair of eyes having checked the changes? Would that be possible with the permissions Github provides?

AFAIK you can lock the master branch and require PRs which are approved and passing tests.

**Comment by obestwalter** [comment]
@The-Compiler - yes you're right. I had a look and there is the possibility to set branches as [protected](https://help.github.com/articles/about-protected-branches/) which would do what is needed for that. 

As a freshly baked maintainer though I would not want to have to bother others with trivial changes and would not want to have to wait for somebody else to approve something I need to get done e.g. as part of a release (like bumping version numbers and modifying the changelog). Also fixing typos and doing simple changes in the docs would be (unnecessarily?) complicated by that. 

On the other hand ... maybe there is an elegant solution for that also. If I look at the commits in tox that document my fumbling around to get the 2.6.0 release into shape I think I should have rather done that on a branch anyway and then squashed that into master instead of cluttering the history with my attempts to get it right. 

I also think that your argument about not everybody wanting to have the commit bit is an important one and should not be underestimated, so it should not be an automatism anyway.

**Comment by obestwalter** [comment]
BTW:  hyper org has a [One Of The Team](https://python-hyper.org/en/latest/one-of-the-team.html) policy which is basically what @flub proposes here and they seem to do well with this.

**Comment by The-Compiler** [comment]
There seems to be quite a difference there though, if I'm reading it correctly:

- The contributor rights aren't given out automatically (there will be people, in my experience quite a few, which *don't* want to be able to do those kind of things, because it brings responsibility they don't want)
- Contributors apparently can't push to `master` directly

**Comment by obestwalter** [comment]
Yes and these are the points that are addressing the concerns voiced here, so everybody wins :)

**Comment by flub** [comment]
So I've updated it to address some concerns raised here.  On the abuse front, I do think that's being overly worried and we can assume people are nice.  If we notice the opposite then we can just revert and stop access.  There is still the bar of having pypi access anyway for security.  And it's not like everyone with commit access now was vetted in any serious way, seems to work :)

So I'm still in favour and would vote for merging this and using it.

**Comment by The-Compiler** [comment]
> If we notice the opposite then we can just revert and stop access.

I'll repeat what I said above - how do we notice? If this is only about PRs I think it's fine, but if everyone can push directly to `master` or `features`, I doubt anyone is going to notice for a loooong time. Or is someone reading the diff each time they pull? :wink:

**Comment by flub** [comment]
You're right, I'm just not terribly concerned about it happening.

I used to read all commits when we where on bitbucket as they ended in my email.  Maybe I should figure out how to enable that again...

**Comment by nicoddemus** [comment]
So @flub has again activated the email notification feature for commits on the main repository.

Do we move this forward or should we decline this one?

**Comment by nicoddemus** [comment]
Guys, should we move this forward or close it?

**Comment by flub** [comment]
I'd still like something along these lines merged, I'll have to re-read the
discussion to see if it should be modified or not.

**Comment by nicoddemus** [comment]
Thanks @flub!

**Comment by nicoddemus** [comment]
@flub gentle ping. 👍

**Comment by flub** [comment]
Ok, I've re-read the whole thing and would still like to commit as-is.  I think the concerns raised here are addressed to reasonable levels.  Yes this can be trivially gamed and people could sneak in malicious code.  But then the system isn't automated either.

So if anyone agrees, feel free to hit the merge button.  If no one does I'll do so in a weeks time (barring objections obviously)

**Comment by RonnyPfannschmidt** [comment]
seems this one went under the bus, i kinda agree and would like to eventually move more towards C4.2 ^^

**Comment by flub** [comment]
\o/

**Pull request merged**